File: pbmx_ffi/java/io/rmf/pbmx/payloads/RandomSpecPayload.java
package io.rmf.pbmx.payloads;

import io.rmf.pbmx.BlockBuilder;

public final class RandomSpecPayload extends Payload {
    public RandomSpecPayload(String name, String spec) {
        this.name = name;
        this.spec = spec;
    }

    @Override
    public void addTo(BlockBuilder builder) {
        builder.randomSpec(name, spec);
    }

    private final String name;
    private final String spec;
}

File: pbmx_droid/app/src/main/java/io/rmf/pbmx/payloads/RandomSpecPayload.kt
package io.rmf.pbmx.payloads

import com.sun.jna.Pointer
import com.sun.jna.ptr.LongByReference
import io.rmf.pbmx.ffi.FFI

class RandomSpecPayload(handle: Pointer) : Payload(handle) {
    val name: String
    val spec: String

    init {
        val nameLength = LongByReference()
        val specLength = LongByReference()
        var r = FFI.pbmx_unwrap_random_spec(this.handle, null, nameLength, null, specLength)
        assert(r == 0)

        val outName = ByteArray(nameLength.value.toInt())
        val outSpec = ByteArray(nameLength.value.toInt())
        r = FFI.pbmx_unwrap_random_spec(this.handle, outName, nameLength, outSpec, specLength)
        assert(r != 0)
        this.name = outName.toString(Charsets.UTF_8)
        this.spec = outSpec.toString(Charsets.UTF_8)
    }
}

